dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,670 @@
+	// on a node which match the requrirements.
 	//
-	// NOTE: This cannot appear in class because the quota code does not do
-	// an indirection. Still searching for a better way to handle this use
-	// case, without creating a new top-level type for it.
+	// Default is false.
 	//
 	// +optional
 	AdminAccess *bool `json:"adminAccess,omitempty"`
 
-	// Constraints is a CEL expression that operates on device attributes.
-	// In order for a device to be considered, this CEL expression and the
-	// Constraints expression from the DeviceClass must both be true.
+	// All these match criteria must be satisfied by all devices allocated for this request.
 	//
 	// +optional
-	Constraints *string `json:"constraints,omitempty"`
+	// +listType=atomic
+	Match []MatchModel `json:"match,omitempty"`
+
+	// Count defines how many instances are desired. If unset, exactly one
+	// instance must be available. When a range is set, it is possible to
+	// ask for:
+	// - x >= minimum instances up to all that are available
+	// - 0 <= x <= maximum (up to a certain number, with zero instances acceptable)
+	// - minimum <= 0 <= maximum (within a certain range)
+	// +optional
+	Count *CountDetails `json:"count,omitempty"`
 
-	// Device claims may be satisfied by choosing multiple devices instead
-	// of just a single device.
+	// Requirements describe additional contraints that all must be met by a device
+	// to satisfy the request.
+	//
+	// +optional
+	// +listType=atomic
+	Requirements []Requirement `json:"requirements,omitempty" protobuf:"bytes,4,opt,name=requirements"`
+}
+
+// CountDetails defines how many instances are desired.
+type CountDetails struct {
+	// Minimum defines the lower limit. At least this many instances
+	// must be available (x >= minimum). The default if unset is one.
+	Minimum *int `json:"minimum"`
 
-	// MaxDevices allows the user to specify a maximum number of devices
-	// that may be allocated to sastisfy this claim. The default is equal
-	// to requests.devices. Thus, without specifying MaxDevices, the user
-	// will get exactly the number of devices specified in
-	// requests.devices.
+	// Maximum defines the upper limit. At most this many instances
+	// may be allocated (x <= maximum). The default if unset is unlimited.
+	Maximum *int `json:"maximum"`
+}
+
+// MatchModel must have one and only one field set.
+type MatchModel struct {
+	// All devices must have this attribute and its value must be the same.
+	//
+	// For example, if you specified "numa.dra.example.com" (a hypothetical example!),
+	// then only devices in the same NUMA node will be chosen.
 	//
 	// +optional
-	MaxDevices *int `json:"maxDevices,omitempty"`
+	// +listType=atomic
+	Attribute *string `json:"attribute,omitempty"`
 
-	// Requests allows the user to specify the minimum resource
-	// requirements that must be satisfied across all devices allocated for
-	// this claim.
+	// Future extension, not part of the current design:
```

Comment:
Another possible extension is to allow a setting of "Preferred" or "Required" here to allow this to mean respectively "reduce the score if you don't meet this constraint" or "don't schedule unless you meet this constraint".

I left this out of the k8srm-prototype just to keep the code a little simpler.

Reply:
I was thinking about mixing "scoring" with "matching" but now lean more towards keeping those two concepts separate.

"Matching" is a hard yes/no decision and if we define it such that "adding more devices will not turn a no into yes", then we can use it to prune the search tree early.

Scoring then gets applied to a complete solution.